### PR TITLE
WIP - MemformerEncoder

### PR DESCRIPTION
**Pull request by bratao**
I´m always trying all your awesome work on transformers. My problem is NER on very large texts, with few examples.

Memformer is the first one so far to converge faster and wield better accuracy than RNN encoders as LSTM, SRU and IndRNN It is ridiculously better than everything else I tested, congratulations @lucidrains 🥳  

I need to use the transformer as a Encoder in my pipeline, to feed a CRF layer. So I modified the code to accept an already embedded input, and to only do the Encode step.

TODO:
- [ ] Support Mask
- [ ] Re-utilize code with Memformer class

Is this within the scope of the project?

### CODE REVIEW

**Comment by lucidrains** [comment]
@bratao wow no way! this works better than routing transformers or any of the other sparse attention techniques? i am surprised

did you already setup MRBP??

i just finished masking in 0.0.6 (but expect bugs)

**Comment by lucidrains** [comment]
@bratao have you ever tried a related work I did in the past? https://github.com/lucidrains/memory-transformer-xl

**Comment by lucidrains** [comment]
@bratao how sure are you that Memformer performs much better than everything you have tried so far? i would be interested in knowing what's your current ranking of solutions (among my repos) for your particular problem?

**Comment by lucidrains** [comment]
@bratao added an encoder only mode `encoder_only = True` that should suit your needs

**Comment by bratao** [comment]
Thank you so much for the encoder @lucidrains 

Yeah, looks crazy right? Maybe I fixed something else on my pipeline, but I will redo the tests now and I can give you a definitive answer. But looking at my older logs this is the testing F-1 for my task.

- 1-order CRF 0.98
- LSTM 0.97
- Memformer 0.97 
- IndRNN 0.95
- nBRC 0.93
- Routing 0.81
- Linear-Transformer  0.33
- Memory-XL 0.27

But take those results with a grain of salt. There is no extensive test on the rights hyper-parms for each network, and my dataset is very small.

**Comment by lucidrains** [comment]
@bratao thank you for the feedback! also give the new Performer a try, if you are willing :D https://github.com/lucidrains/performer-pytorch

**Comment by lucidrains** [comment]
@bratao do let me know if you continue to find success with memformer, and we can push the boundaries of this architecture

**Comment by lucidrains** [comment]
@bratao hey! i saw you promoting performer on reddit lol

have you tried Performer on your task? would be very curious to know what your results were!

**Comment by bratao** [comment]
@lucidrains yeah, I get super excited to try newer architectures, and it helps me understand the algorithms better. 

I did not tried your version of performer yet, just the fast-transformer with FAVOR kernel, that is not exactly performer. However this week I´m planning to retest all your recommended transformers architectures on a T4 GPU with 16GB. As soon as I have the results I send to you.

One thing that I miss about your line of transformers is the possibility of using an already embedded input. I always have to comment  nn.Embedding(num_tokens, dim). It is just a minor thing, but the way the huggingface transformers do is neat (https://github.com/huggingface/transformers/blob/f744b81572e533af5a8469c2fba661c5972f2b66/src/transformers/modeling_reformer.py#L2007)

**Comment by lucidrains** [comment]
@bratao I'll see what I can do about passing in inputs as embeddings :smile:
